oss-internship-2020/libarchive/examples/sandbox.h

```diff
@@ -0,0 +1,144 @@
+// map "/output" with the basename of the archive. This way, the program can
+// create the file without having access to anything else.
+class SapiLibarchiveSandboxCreate : public LibarchiveSandbox {
+ public:
+  explicit SapiLibarchiveSandboxCreate(const std::vector<std::string>& files,
+                                       absl::string_view archive_path)
+      : files_(files), archive_path_(archive_path) {}
+
+ private:
+  std::unique_ptr<sandbox2::Policy> ModifyPolicy(
+      sandbox2::PolicyBuilder*) override {
+    sandbox2::PolicyBuilder policy =
+        sandbox2::PolicyBuilder()
+            .AddDirectoryAt(archive_path_, "/output", false)
+            .AllowRead()
+            .AllowWrite()
+            .AllowOpen()
+            .AllowSystemMalloc()
+            .AllowGetIDs()
+            .AllowSafeFcntl()
+            .AllowStat()
+            .AllowExit()
+            .AllowSyscalls({
+                __NR_futex,
+                __NR_lseek,
+                __NR_close,
+                __NR_gettid,
+                __NR_umask,
+                __NR_utimensat,
+                __NR_unlink,
+                __NR_mkdir,
+                __NR_fstatfs,
+                __NR_socket,
+                __NR_connect,
+                __NR_ioctl,
+                __NR_flistxattr,
+                __NR_recvmsg,
+                __NR_getdents64,
+            });
+
+    // We check whether the entry is a file or a directory.
+    for (const auto& i : files_) {
+      struct stat s;
+      stat(i.c_str(), &s);
+      if (S_ISDIR(s.st_mode)) {
+        policy = policy.AddDirectory(i);
+      } else {
+        policy = policy.AddFile(i);
+      }
+    }
+
+    return policy.BuildOrDie();
+  }
+
+  const std::vector<std::string> files_;
+  absl::string_view archive_path_;
+};
+
+// When an archive is extracted, the generated files/directories will be placed
+// relative to the current working directory. In order to add permissions to
+// this we create a temporary directory at every extraction. Then, we change the
+// directory of the sandboxed process to that directory and map it to the
+// current "real" working directory. This way the contents of the archived will
+// pe placed correctly without offering additional permission.
+class SapiLibarchiveSandboxExtract : public LibarchiveSandbox {
+ public:
+  explicit SapiLibarchiveSandboxExtract(absl::string_view archive_path,
+                                        const int do_extract,
+                                        absl::string_view tmp_dir)
+      : archive_path_(archive_path),
+        do_extract_(do_extract),
+        tmp_dir_(tmp_dir) {}
+
+ private:
+  virtual void ModifyExecutor(sandbox2::Executor* executor) override {
+    // If the user only wants to list the entries in the archive, we do
+    // not need to worry about changing directories;
+    if (do_extract_) {
+      executor = &executor->set_cwd(std::string(tmp_dir_));
```

Comment:
`executor = &executor` not needed, see https://github.com/google/sandboxed-api/blob/master/sandboxed_api/sandbox2/executor.h#L92 - it allows for function chaining, but can be used without.